rust/connlib/tunnel/src/io.rs

```diff
@@ -186,9 +221,69 @@ pub enum DnsQueryError {
     TooManyQueries,
 }
 
-fn create_resolvers(
+/// Exactly the same as the [TokioRuntimeProvider](hickory_resolver::name_server::TokioRuntimeProvider) but sockets are protected when created
+#[derive(Clone)]
+struct ProtectedTokioRuntimeProvider<S> {
+    handle: TokioHandle,
+    socket_protect: S,
```

Comment:
You can make this a trait-object.

Reply:
```suggestion
    socket_protect: Box<dyn RuntimeSocketProtector>,
```

Reply:
The problem is that `RuntimeProvider` needs to implement `Clone` for that `RuntimeSocketProtector` needs to implement `Clone` or `socket_protect` needs to be an `Arc`, which we can do, I just wanted to avoid it since I think seeing an `Arc` here can be a bit weird.

Reply:
I'd say that can easily enough be explained in a comment if it buys us removal of the type-parameter! :)